File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {

        Deque<Character> stack = new ArrayDeque<>();

        for (char currentChar : s.toCharArray()) {
            if (currentChar == '(' || currentChar == '{' || currentChar == '[') {
                stack.push(currentChar);
            }

            else {
                if (stack.isEmpty() || !isMatchingPair(stack.peek(), currentChar)) {
                    return false;
                }
                stack.pop();
            }
        }

        return stack.isEmpty();
    }

    private boolean isMatchingPair(char openingBracket, char closingBracket) {
        return (openingBracket == '(' && closingBracket == ')') ||
               (openingBracket == '{' && closingBracket == '}') ||
               (openingBracket == '[' && closingBracket == ']');
    }
}

File: 20-valid-parentheses/valid-parentheses.go
func isValid(s string) bool {
    stack := []rune{}
	matching := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}
	for _, char := range s {
		switch char {
			case '(', '{', '[':
				stack = append(stack, char) // push
			case ')', '}', ']':
				if len(stack) == 0 || stack[len(stack)-1] != matching[char] {
					return false
				}
				stack = stack[:len(stack)-1] // pop
		}
	}
	return len(stack) == 0
}